ginga/rv/plugins/WBrowser.py

```diff
@@ -0,0 +1,119 @@
+#
+# WBrowser.py -- Web Browser plugin for Ginga
+#
+# This is open-source software licensed under a BSD license.
+# Please see the file LICENSE.txt for details.
+#
+"""Browse help for Ginga."""
+from __future__ import absolute_import, division, print_function
+
+# GINGA
+from ginga.GingaPlugin import GlobalPlugin
+from ginga.gw import Widgets
+
+__all__ = []
+
+
+class WBrowser(GlobalPlugin):
+    """
+    WBrowser
+    ========
+    Web Browser plugin for Ginga.
+
+    Plugin Type: Global
+    -------------------
+    WBrowser is a global plugin. Only one instance can be opened.
+
+    This global plugin is used to browse help pages for Ginga.
+    When available, a local help page is displayed.
+
+    """
+    def __init__(self, fv):
+        # superclass defines some variables for us, like logger
+        super(WBrowser, self).__init__(fv)
+
+    def build_gui(self, container):
+        if not Widgets.has_webkit:
+            self.browser = Widgets.Label(
+                "Please install the python-webkit package to enable "
+                "this plugin")
+        else:
+            self.browser = Widgets.WebView()
+
+        sw = Widgets.ScrollArea()
+        sw.set_widget(self.browser)
+
+        container.add_widget(sw, stretch=1)
+        sw.show()
+
+        self.entry = Widgets.TextEntry()
+        container.add_widget(self.entry, stretch=0)
+        self.entry.add_callback('activated', lambda w: self.browse_cb())
+
+        btns = Widgets.HBox()
+        btns.set_border_width(4)
+        btns.set_spacing(3)
+
+        btn = Widgets.Button('Close')
+        btn.add_callback('activated', lambda w: self.close())
+        btns.add_widget(btn, stretch=0)
+        btns.add_widget(Widgets.Label(''), stretch=1)
+        container.add_widget(btns, stretch=0)
+
+        self.show_help()
+        self.gui_up = True
+
+    def show_help(self, plugin=None, no_url_callback=None):
+        """See `~ginga.GingaPlugin` for usage of optional keywords."""
+        if not Widgets.has_webkit:
+            return
+
+        from ginga.doc.download_doc import get_doc
+        self.fv.assert_gui_thread()
+
+        def _dl_indicator(count, blksize, totalsize):
+            pct = (count * blksize) / totalsize
+            msg = 'Downloading: {:.2%} complete'.format(pct)
+            self.fv.gui_do(self.entry.set_text, msg)
+
+        helpurl = get_doc(logger=self.logger, plugin=plugin,
```

Comment:
You have to split the work units so that they are chained from thread to thread via functions and not inline anymore.  So this would become:
```python
    self.fv.nongui_do(self._download_doc, plugin, _dl_indicator, no_url_callback)

def _download_doc(self, plugin, fn_indicator, no_url_callback):
    self.fv.assert_nongui_thread()

    # this can block as long as it takes without blocking the UI
    try:
        helpurl = get_doc(logger=self.logger, plugin=plugin, reporthook=fn_indicator)

    except Exception as e:
        self.fv.gui_do(self.fv.show_error, "Error downloading documentation: %s" % (str(e)))

    self.fv.gui_do(self._load_doc, helpurl, no_url_callback)

def _load_doc(self, helpurl, no_url_callback):
    self.fv.assert_gui_thread()
    if helpurl is None:
￼        self.fv.entry.set_text('')
￼        if no_url_callback is None:
              self.fv.show_error("Couldn't load web page %s": (helpurl))
￼        else:
              no_url_callback()
￼    else:
￼         self.browse(helpurl)
```
Yes it breaks up the flow, but unfortunately there is no easy way to have a single code stream switch between threads.